File: electron-app/mcp_server.py
# ============================================================
# mcp_server.py - 通用 MCP 服务器（完全放开版）
# 没有任何路径限制，AI 可以读写任何文件
# 
# 在 Dify 中添加 MCP 工具时填入：
#   http://host.docker.internal:18889
#
# 如何添加新工具？
# 1. 写一个函数，用 @register_tool 装饰
# 2. 重启服务器
# 搞定。
# ============================================================
import sys
sys.stdout.reconfigure(line_buffering=True)
import os
import sys
import json
from http.server import HTTPServer, BaseHTTPRequestHandler

if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8')
    sys.stderr.reconfigure(encoding='utf-8')

# ============================================================
# 工具注册表
# ============================================================

TOOL_REGISTRY = {}

def register_tool(name, description, input_schema):
    def decorator(func):
        TOOL_REGISTRY[name] = {
            "fn": func,
            "desc": description,
            "schema": input_schema
        }
        return func
    return decorator


# ============================================================
# 工具函数定义区
# ============================================================

# ----- 工具 1：列出目录 -----
@register_tool(
    name="list_files",
    description="列出指定目录中的文件和文件夹",
    input_schema={
        "type": "object",
        "properties": {
            "path": {
                "type": "string",
                "description": "目录路径，默认为桌面"
            }
        }
    }
)
def handle_list_files(args):
    path = args.get("path", "D:/")
    try:
        items = os.listdir(path)
        return f"成功！路径 {path} 下共有 {len(items)} 项：\n" + "\n".join(items[:50])
    except Exception as e:
        return f"失败：{type(e).__name__} - {str(e)}"

# ----- 工具 2：读取文件 -----
@register_tool(
    name="read_file",
    description="读取文件内容",
    input_schema={
        "type": "object",
        "properties": {
            "path": {
                "type": "string",
                "description": "文件完整路径"
            }
        },
        "required": ["path"]
    }
)
def handle_read_file(args):
    path = args.get("path", "")
    if not path:
        return "错误：缺少路径参数"
    try:
        with open(path, 'r', encoding='utf-8') as f:
            content = f.read()
        return "文件：" + path + "\n\n" + content
    except Exception as e:
        return "错误：" + str(e)


# ----- 工具 3：写入文件 -----
@register_tool(
    name="write_file",
    description="写入文件内容",
    input_schema={
        "type": "object",
        "properties": {
            "path": {
                "type": "string",
                "description": "文件完整路径"
            },
            "content": {
                "type": "string",
                "description": "要写入的内容"
            }
        },
        "required": ["path", "content"]
    }
)
def handle_write_file(args):
    path = args.get("path", "")
    content = args.get("content", "")
    if not path:
        return "错误：缺少路径参数"
    try:
        with open(path, 'w', encoding='utf-8') as f:
            f.write(content)
        return "已写入文件：" + path + "（" + str(len(content)) + " 字符）"
    except Exception as e:
        return "错误：" + str(e)


# ----- 工具 4：搜索文件 -----
@register_tool(
    name="search_files",
    description="按文件名搜索文件",
    input_schema={
        "type": "object",
        "properties": {
            "keyword": {
                "type": "string",
                "description": "搜索关键词"
            },
            "path": {
                "type": "string",
                "description": "搜索起始目录，默认为桌面"
            }
        },
        "required": ["keyword"]
    }
)
def handle_search_files(args):
    keyword = args.get("keyword", "")
    path = args.get("path", os.path.expanduser("~\\Desktop"))
    if not keyword:
        return "错误：缺少搜索关键词"
    try:
        results = []
        for root, dirs, files in os.walk(path):
            depth = root.replace(path, '').count(os.sep)
            if depth > 5:
                continue
            for file in files:
                if keyword.lower() in file.lower():
                    full_path = os.path.join(root, file)
                    size = os.path.getsize(full_path)
                    results.append(f"{full_path} ({size} 字节)")
        if results:
            return "搜索「" + keyword + "」找到 " + str(len(results)) + " 个结果：\n\n" + "\n".join(results)
        else:
            return "未找到包含「" + keyword + "」的文件"
    except Exception as e:
        return "错误：" + str(e)


# ============================================================
# MCP 协议层（不需要修改）
# ============================================================

class MCPHandler(BaseHTTPRequestHandler):
    
    def do_GET(self):
        if self.path in ['/health', '/mcp/health']:
            self.send_json(200, {"status": "ok", "service": "mcp-server", "tools": list(TOOL_REGISTRY.keys())})
            return
        self.send_json(404, {"error": "not found"})
    
    def do_POST(self):
        content_length = int(self.headers.get('Content-Length', 0))
        body = self.rfile.read(content_length) if content_length > 0 else b'{}'
        
        try:
            req_json = json.loads(body)
        except:
            req_json = {}
        
        # 打印收到的请求（调试用）
        print("[收到]", json.dumps(req_json, ensure_ascii=False, indent=2))
        
        method = req_json.get('method', '')
        req_id = req_json.get('id', None)
        
        # 1. 初始化
        if method == 'initialize':
            client_version = req_json.get('params', {}).get('protocolVersion', '2025-06-18')
            self.send_json(200, {
                "jsonrpc": "2.0",
                "result": {
                    "protocolVersion": client_version,
                    "capabilities": {"tools": {}},
                    "serverInfo": {"name": "mcp-server", "version": "1.0.0"}
                },
                "id": req_id
            })
        
        # 2. 初始化完成通知
        elif method == 'notifications/initialized':
            self.send_response(202)
            self.send_header('Content-Type', 'application/json')
            self.end_headers()
            self.wfile.write(b'{}')
        
        # 3. 列出工具
        elif method == 'tools/list':
            tools = []
            for name, info in TOOL_REGISTRY.items():
                tools.append({
                    "name": name,
                    "description": info["desc"],
                    "inputSchema": info["schema"]
                })
            self.send_json(200, {
                "jsonrpc": "2.0",
                "result": {"tools": tools},
                "id": req_id
            })
        
        # 4. 调用工具
        elif method == 'tools/call':
            params = req_json.get('params', {})
            name = params.get('name', '')
            arguments = params.get('arguments', {})
            
            if name in TOOL_REGISTRY:
                try:
                    result_text = TOOL_REGISTRY[name]["fn"](arguments)
                except Exception as e:
                    result_text = "执行错误：" + str(e)
            else:
                result_text = "未知工具：" + name
            
            self.send_json(200, {
                "jsonrpc": "2.0",
                "result": {
                    "content": [{"type": "text", "text": result_text}]
                },
                "id": req_id
            })
        
        else:
            self.send_json(200, {
                "jsonrpc": "2.0",
                "result": {"content": [{"type": "text", "text": f"未知方法: {method}"}]},
                "id": req_id
            })
    
    def send_json(self, status, data):
        self.send_response(status)
        self.send_header('Content-Type', 'application/json; charset=utf-8')
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Headers', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.end_headers()
        self.wfile.write(json.dumps(data, ensure_ascii=False).encode('utf-8'))
    
    def do_OPTIONS(self):
        self.send_response(200)
        self.send_header('Access-Control-Allow-Origin', '*')
        self.send_header('Access-Control-Allow-Headers', '*')
        self.send_header('Access-Control-Allow-Methods', 'GET, POST, OPTIONS')
        self.end_headers()
    
    def log_message(self, format, *args):
        pass


def main():
    HOST = "0.0.0.0"
    PORT = 18889
    
    server = HTTPServer((HOST, PORT), MCPHandler)
    
    print("=" * 50)
    print("  通用 MCP 服务器（完全放开版）")
    print(f"  地址: http://{HOST}:{PORT}")
    print(f"  在 Dify 中填入: http://host.docker.internal:{PORT}")
    print("=" * 50)
    print(f"\n  已注册 {len(TOOL_REGISTRY)} 个工具:")
    for name in TOOL_REGISTRY:
        print(f"    - {name}: {TOOL_REGISTRY[name]['desc']}")
    print(f"\n  按 Ctrl+C 停止")
    
    server = HTTPServer((HOST, PORT), MCPHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        print("\n正在关闭服务器...")
        server.server_close()
        print("服务器已停止。")


if __name__ == "__main__":
    main()

File: electron-app/mcp-server.ts
import { FastMCP } from "fastmcp";
import { z } from "zod";
import fs from "fs/promises";
import path from "path";
import os from "os";

// 创建 FastMCP 服务器
const server = new FastMCP({
  name: "Filesystem Server (TypeScript)",
  version: "1.0.0",
});

// ---------- 辅助函数 ----------
function getDefaultPath(): string {
  // Windows 桌面路径
  return path.join(os.homedir(), "Desktop");
}

// 安全获取文件大小，避免权限错误
async function safeGetSize(filePath: string): Promise<number | null> {
  try {
    const stat = await fs.stat(filePath);
    return stat.size;
  } catch {
    return null;
  }
}

// ---------- 工具1: list_files ----------
server.addTool({
  name: "list_files",
  description: "列出指定目录中的文件和文件夹",
  parameters: z.object({
    path: z.string().optional().describe("目录路径，默认为桌面"),
  }),
  execute: async (args) => {
    const targetPath = args.path || getDefaultPath();
    try {
      const items = await fs.readdir(targetPath);
      const dirs: string[] = [];
      const files: string[] = [];

      for (const item of items) {
        const fullPath = path.join(targetPath, item);
        try {
          const stat = await fs.stat(fullPath);
          if (stat.isDirectory()) {
            dirs.push(item);
          } else {
            const size = await safeGetSize(fullPath);
            files.push(`${item} (${size !== null ? size + " 字节" : "大小未知"})`);
          }
        } catch (err) {
          // 权限问题：标记为无法访问
          dirs.push(`${item} (系统保护，无法读取详情)`);
        }
      }

      const result = [
        `目录：${targetPath}`,
        "",
        "文件夹：",
        ...dirs.map(d => "  " + d),
        "",
        "文件：",
        ...files.map(f => "  " + f),
      ].join("\n");
      return result;
    } catch (err: any) {
      return `错误：${err.message || err}`;
    }
  },
});

// ---------- 工具2: read_file ----------
server.addTool({
  name: "read_file",
  description: "读取文件内容",
  parameters: z.object({
    path: z.string().describe("文件完整路径"),
  }),
  execute: async (args) => {
    try {
      const content = await fs.readFile(args.path, "utf-8");
      return `文件：${args.path}\n\n${content}`;
    } catch (err: any) {
      return `错误：${err.message || err}`;
    }
  },
});

// ---------- 工具3: write_file ----------
server.addTool({
  name: "write_file",
  description: "写入文件内容",
  parameters: z.object({
    path: z.string().describe("文件完整路径"),
    content: z.string().describe("要写入的内容"),
  }),
  execute: async (args) => {
    try {
      // 确保目录存在
      const dir = path.dirname(args.path);
      await fs.mkdir(dir, { recursive: true });
      await fs.writeFile(args.path, args.content, "utf-8");
      return `已写入文件：${args.path}（${args.content.length} 字符）`;
    } catch (err: any) {
      return `错误：${err.message || err}`;
    }
  },
});

// ---------- 工具4: search_files ----------
server.addTool({
  name: "search_files",
  description: "按文件名搜索文件",
  parameters: z.object({
    keyword: z.string().describe("搜索关键词"),
    path: z.string().optional().describe("搜索起始目录，默认为桌面"),
  }),
  execute: async (args) => {
    const targetPath = args.path || getDefaultPath();
    const keyword = args.keyword;
    if (!keyword) return "错误：缺少搜索关键词";

    const results: string[] = [];
    const maxDepth = 5;

    async function walk(dir: string, depth: number) {
      if (depth > maxDepth) return;
      try {
        const items = await fs.readdir(dir);
        for (const item of items) {
          const fullPath = path.join(dir, item);
          try {
            const stat = await fs.stat(fullPath);
            if (stat.isDirectory()) {
              await walk(fullPath, depth + 1);
            } else {
              if (item.toLowerCase().includes(keyword.toLowerCase())) {
                const size = await safeGetSize(fullPath);
                results.push(`${fullPath} (${size !== null ? size + " 字节" : "大小未知"})`);
              }
            }
          } catch {
            // 忽略权限错误继续
          }
        }
      } catch {
        // 忽略无法读取的目录
      }
    }

    await walk(targetPath, 0);

    if (results.length > 0) {
      return `搜索「${keyword}」找到 ${results.length} 个结果：\n\n${results.join("\n")}`;
    } else {
      return `未找到包含「${keyword}」的文件`;
    }
  },
});

// ---------- 启动 HTTP 服务 ----------
async function main() {
  // 使用 HTTP 传输，监听所有地址，端口 18889
  await server.start({
    transportType: "httpStream",
    httpStream: {
      port: 18889,
      host: "0.0.0.0",
    },
  });
  console.log(`✅ MCP 服务器已启动: http://0.0.0.0:18889`);
  console.log(`   在 Dify 中配置: http://host.docker.internal:18889`);
}

main().catch((err) => {
  console.error("启动失败:", err);
  process.exit(1);
});